Extract get samples

dlebauer opened the pull request:
This replaces and fixes #187, #99. This extracts a get.parameter.samples function from run.write.configs in order to make it more modular. I've set this to run at the end of the meta-analysis, saving `samples.Rdata` to the `settings$outdir` that it is not necessary to run the write configs in order to extract the parameter samples that will be used for the runs. 

`run.write.configs` now loads the `samples.Rdata` file, and is considerably cleaned up.

This could be a separate step in the workflow, but requiring this would break old workflows. It is just as easy to embed the saving into run.meta.analysis and then load the samples in run.write.configs.

dlebauer commented:
still running tests ...

mdietze commented:
@dlebauer I really don't follow the logic of moving the parameter sampling for the sensitivity and uncertainty analyses (which fundamentally should be part of the uncertainty module) into the meta analysis module (and embedded within the main MA function) which is doing something that is conceptually completely different. This would also break the data assimilation workflows, which do sensitivity and ensemble analyses on the post-DA posteriors, not the post-MA posteriors. 

If run.write.configs is going to get a major overhaul to move parameter sampling out of it, you might think about going the whole 9 yards and removing all 'sensitivity' and 'uncertainty' specific bits to make this a function that is generically useful for writing configs? Right now both the parameter and state data assimilation modules have to reinvent large chunks of write configs because they're trying to start runs for purposes other than SA or EA. Would be easier to maintain/update if we didn't have similar redundant code sprinkled across a bunch of modules.

dlebauer commented:
> don't follow the logic of moving the parameter sampling for the sensitivity and uncertainty analyses (which fundamentally should be part of the uncertainty module) into the meta analysis 

The idea was that it is useful to be able to extract these values before writing configuration files, specifically to be able to provide a set of parameterizations that can be used independently of PEcAn. For this purpose, it is sufficient that the function `get.parameter.samples` works independently of `run.write.configs`, and hopefully by adding it back to the beginning of run.write.configs does not break the other uses of this function.

I've moved the `get.parameter.samples function to the uncertainty package, and it is now run at the beginning of the `run.write.configs` function rather than at the end of `run.meta.analysis`. 

> This would also break the data assimilation workflows, 

Is there a way to incorporate these  workflows into the integration tests? Either as package tests, vignettes that are executed by `R CMD build --check`, or as another file like `workflow.R`?

> you might think about going the whole 9 yards and removing all 'sensitivity' and 'uncertainty' specific bits 

This is a very good idea but not a priority - now that I've implemented this refactoring for a third time (previously #187, #99), I am looking forward to merging into master before anyone else merges an incompatible change.
